### PR TITLE
 on the fly reprojection for url, api, xapi and full text search

### PR DESCRIPTION
**Pull request by gilliM**

### CODE REVIEW

File: core/examples/fulltextsearch.js [hunk 1]
```diff
@@ -20,7 +20,6 @@ Ext.onReady(function() {
             coordsRecenterZoom: 11,
             widgetOptions: {
               tooltipEl: 'search-tip',
-              projectionCodes: [21781, 4326],
```
Comment (asaunier) [inline core/examples/fulltextsearch.js]: If we are really sure that we want to move ``projectionCodes`` config from the widget to the plugin, you need to add an entry in CGXP's changelog.

File: core/src/script/CGXP/tools/autoProject.js [hunk 2]
```diff
@@ -0,0 +1,91 @@
+/**
```
Comment (asaunier) [inline core/src/script/CGXP/tools/autoProject.js]: This the class is named ``OpenLayers.AutoProjection`` you should consider renaming the file to ``AutoProjection.js`` and move it to ``openlayers.addins/``, don't you think?

Comment (asaunier) [inline core/src/script/CGXP/tools/autoProject.js]: But why name it ``OpenLayers.AutoProjection``? Isn't it some CGXP specific code?

Comment (gilliM) [inline core/src/script/CGXP/tools/autoProject.js]: It has to be available for the Api. So it has to be OpenLayers.

On Fri, Feb 6, 2015 at 9:09 AM, Alex Saunier <notifications@github.com>
wrote:

> In core/src/script/CGXP/tools/autoProject.js
> <https://github.com/camptocamp/cgxp/pull/909#discussion_r24227332>:
>
> > @@ -0,0 +1,91 @@
> > +/**
>
> But why name it OpenLayers.AutoProjection? Isn't it some CGXP specific
> code?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/camptocamp/cgxp/pull/909/files#r24227332>.
>



-- 

*camp**to**camp*
INNOVATIVE SOLUTIONS
BY OPEN SOURCE EXPERTS

*Gillian Milani*

Ingénieur en environnement
Stagiaire

+41 21 619 10 10 (central)
+41 21 619 10 48 (direct)
*Camptocamp SA*
Quartier de l'Innovation EPFL, PSE-A
1015 Lausanne
Suisse

File: core/src/script/CGXP/widgets/FullTextSearch.js [hunk 3]
```diff
@@ -138,17 +126,7 @@ cgxp.FullTextSearch = Ext.extend(Ext.Panel, {
...
-            this.projectionCodes = [this.map.getProjection()];
-        }
-        for (var i = 0, len = this.projectionCodes.length, code; i < len; i++) {
-            code = String(this.projectionCodes[i]).toUpperCase();
-            if (code.substr(0, 5) != "EPSG:") {
-                code = "EPSG:" + code;
-            }
-            this.projections[code] = new OpenLayers.Projection(code);
-        }
+        this.autoProjection = new OpenLayers.AutoProjection(this);
```
Comment (asaunier) [inline core/src/script/CGXP/widgets/FullTextSearch.js]: Using this class requires that you add an @include statement at the beginning of the file.

File: core/src/script/CGXP/tools/autoProject.js [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+           this.isEnabled = true;
+        } else { 
+           this.isEnabled = false;
+        }
+    },
+
+    /** private: property[projections]
+     *  List of OpenLayers.Projection object
+     *  used to reproject some coordinates. It's based
+     *  on projectionCodes
```
Comment (asaunier) [inline core/src/script/CGXP/tools/autoProject.js]: I don't think you need to specify projectionCodes here, it is already done in the comments of buildProjectionList().

File: core/src/script/CGXP/api/Map.js [hunk 5]
```diff
@@ -40,6 +41,26 @@
  *  class = Map
  */
 
+/** api: autoprojection - example
+ *  
+ *  Sample code showing how to configure the mappanel for autoprojection.
```
Comment (asaunier) [inline core/src/script/CGXP/api/Map.js]: What is "autoprojection"?

File: core/src/script/CGXP/tools/autoProject.js [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+     *  and fill the projections dictionary with Openlayers.Projection
+     *  objects.
+     */
+    buildProjectionList: function(config){
+        this.projections = [];
+        if (!config.projectionCodes) {
+            return;
+        }
+        for (var i = 0, len = config.projectionCodes.length, code; i < len; i++)
+            {
```
Comment (asaunier) [inline core/src/script/CGXP/tools/autoProject.js]: "{" should be at the end of the previous line.

File: core/src/script/CGXP/tools/autoProject.js (hunk 4 above)
Comment (gilliM) [inline core/src/script/CGXP/tools/autoProject.js]: mh... projectionCodes is a parameter from the config. buildProjectionList transform it to "projections". I'm not sure of your remark...

File: core/src/script/CGXP/tools/autoProject.js [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+        }
+        return;
+    },
+
+    /** private: method[tryProjection]
+     *  :arg point: ``Array`` point to project
+     *  :arg map: ``Object`` the map object
+     *
+     *  Project the point using the projections list
+     *  and find the first projection for which the it
```
Comment (asaunier) [inline core/src/script/CGXP/tools/autoProject.js]: This sentence does not make sense ("for which the it falls")

File: core/src/script/CGXP/widgets/MapPanel.js [hunk 8]
```diff
@@ -60,6 +60,29 @@ Ext.namespace("cgxp");
  *   - Example: ``&map_tooltip=sometext``
  *
  */
+
+/** api: autoprojection - example
+ *  
+ *  Sample code showing how to configure the mappanel for autoprojection.
```
Comment (asaunier) [inline core/src/script/CGXP/widgets/MapPanel.js]: Ditto here: what is autoproject?

File: core/src/script/CGXP/widgets/MapPanel.js [hunk 9]
```diff
@@ -60,6 +60,29 @@ Ext.namespace("cgxp");
...
+ *  Here, the application is in 21781, the restricted extent corresponds to 
+ *  Swiss boundaries, and supported projection are 21781 and 2056.
+ *
+ *  The autoprojection will be enabled for URL and FTS.
```
Comment (asaunier) [inline core/src/script/CGXP/widgets/MapPanel.js]: FTS = Full Text Search

File: core/src/script/CGXP/tools/autoProject.js (hunk 2 above)
Comment (asaunier) [inline core/src/script/CGXP/tools/autoProject.js]: OK!

File: core/examples/fulltextsearch.js (hunk 1 above)
Comment (gilliM) [inline core/examples/fulltextsearch.js]: Do i need to do a PR in demo_geomapfish?

Comment (asaunier) [inline core/examples/fulltextsearch.js]: Yes, else the demo will be broken next time CGXP is updated, isn't it?

File: core/src/script/CGXP/api/Map.js [hunk 10]
```diff
@@ -217,6 +243,13 @@ cgxp.api.Map.prototype = {
     onViewerReady: function(viewer) {
         var i;
         this.map = viewer.mapPanel.map;
+        this.autoProjection = new OpenLayers.AutoProjection(viewer.map); 
+
+        if (typeof this.userConfig.center != 'undefined') {
```
Comment (sbrunner) [inline core/src/script/CGXP/api/Map.js]: => this.userConfig.center !== undefined

File: core/src/script/CGXP/api/Map.js [hunk 11]
```diff
@@ -217,6 +243,13 @@ cgxp.api.Map.prototype = {
...
+            var zoom = (typeof this.userConfig.zoom == 'undefined' ? 10 :
```
Comment (sbrunner) [inline core/src/script/CGXP/api/Map.js]: idem

File: core/src/script/CGXP/api/Map.js [hunk 12]
```diff
@@ -249,8 +284,10 @@ cgxp.api.Map.prototype = {
         }
 
         OpenLayers.Util.extend(config, this.userConfig);
-
         this.map = new OpenLayers.Map(config);
+        if ( typeof(config.center) != 'undefined') {
```
Comment (sbrunner) [inline core/src/script/CGXP/api/Map.js]: idem

File: core/src/script/CGXP/api/Map.js [hunk 13]
```diff
@@ -433,6 +476,10 @@ cgxp.api.Map.prototype = {
             return;
         }
         options = options || {};
+        if (typeof options.position != 'undefined'){
```
Comment (sbrunner) [inline core/src/script/CGXP/api/Map.js]: idem

File: openlayers.addins/AutoProjection/lib/OpenLayers/Util/AutoProjection.js [hunk 14]
```diff
@@ -0,0 +1,86 @@
...
+    tryProjection: function(point, map) {
+        if (this.projections.length === 0) {
+            this.projections = [map.projection];
+        }
+        for (var i=0; i < this.projections.length; ++i) {
+            var projection = this.projections[i];
+            var position = new OpenLayers.LonLat(point[0], point[1]);
+            position.transform(projection, map.projection);
+            if (map.restrictedExtent.containsLonLat(position)) {
+                return [position.lon, position.lat];
```
Comment (sbrunner) [inline openlayers.addins/AutoProjection/lib/OpenLayers/Util/AutoProjection.js]: We nearly always needs a OpenLayers.LonLat, than we should return it !

File: core/src/script/CGXP/api/Map.js [hunk 15]
```diff
@@ -357,6 +394,12 @@ cgxp.api.Map.prototype = {
                 OpenLayers.Function.bind(this.recenter, this, center, zoom));
             return;
         }
+        center = this.autoProjection.tryProjection(center, this.map);
+        if (center == null){
+            center = []
+            center[0] = null;
+            center[1] = null;
+        }
         this.map.setCenter(new OpenLayers.LonLat(center[0], center[1]), zoom);
```
Comment (sbrunner) [inline core/src/script/CGXP/api/Map.js]: => 
```
var centerLonLat = this.autoProjection.tryProjection(center, this.map);
if (centerLonLat === undefined) {
  centerLonLat = 
this.map.setCenter(new OpenLayers.LonLat(center[0], center[1]);
}
this.map.setCenter(centerLonLat, zoom);
}

File: core/src/script/CGXP/widgets/FullTextSearch.js [hunk 16]
```diff
@@ -116,18 +120,6 @@ cgxp.FullTextSearch = Ext.extend(Ext.Panel, {
      */
     limits: {},
 
-    /** api: config[projectionCodes]
-     *  ``Array``
-     *  List of EPSG codes of projections that should be used when trying to
-     *  recenter on coordinates. Leftmost projections are used preferably.
-     *  Default is current map projection.
-     */
-    projectionCodes: null,
```
Comment (sbrunner) [inline core/src/script/CGXP/widgets/FullTextSearch.js]: We should add a cgĥangelog message about it :-)

File: core/src/script/CGXP/widgets/FullTextSearch.js [hunk 17]
```diff
@@ -177,23 +159,10 @@ cgxp.FullTextSearch = Ext.extend(Ext.Panel, {
...
-                    return false;
-                }.createDelegate(this);
-
-                for (var projection in this.projections) {
-                    if (tryProjection(left, right, projection) ||
-                        tryProjection(right, left, projection)) {
-                        break;
-                    }
+                this.position = this.autoProjection.tryProjection([left, right], map); 
+                if (this.position == null){
```
Comment (sbrunner) [inline core/src/script/CGXP/widgets/FullTextSearch.js]: missing space between ) {
You should use the operator === to test with null

File: core/src/script/CGXP/widgets/MapPanel.js [hunk 18]
```diff
@@ -73,6 +102,15 @@ cgxp.MapPanel = Ext.extend(GeoExt.MapPanel, {
      */
     crosshairStyle: {},
 
+    /** api: config[projectionCodes]
+    * ``Array``
+    *
+    * List of EPSG codes of projections that should be used when trying to
+    * recenter on coordinates. Leftmost projections are used preferably.
+    * Default is current map projection.
+    */
```
Comment (sbrunner) [inline core/src/script/CGXP/widgets/MapPanel.js]: indent issue

File: core/src/script/CGXP/widgets/MapPanel.js [hunk 19]
```diff
@@ -148,6 +188,16 @@ cgxp.MapPanel = Ext.extend(GeoExt.MapPanel, {
      */
     applyState: function(state) {
         this.initialState = state;
+        var newCenter = this.autoProjection.tryProjection(
+                                 [state.x, state.y],
+                                 this.map);
+        if (newCenter == null){
```
Comment (sbrunner) [inline core/src/script/CGXP/widgets/MapPanel.js]: Missing space between ) and {

File: openlayers.addins/AutoProjection/lib/OpenLayers/Util/AutoProjection.js [hunk 20]
```diff
@@ -0,0 +1,86 @@
...
+
+/**
+* Namespace: AutoProjection
+*/
+
+/**
+* @requires OpenLayers/BaseTypes/Class.js
+* @requires OpenLayers/Util.js
+* @include OpenLayers/Map.js
+*/
```
Comment (sbrunner) [inline openlayers.addins/AutoProjection/lib/OpenLayers/Util/AutoProjection.js]: Indent issues

File: openlayers.addins/AutoProjection/lib/OpenLayers/Util/AutoProjection.js [hunk 21]
```diff
@@ -0,0 +1,86 @@
...
+
+    /** private: method[tryProjection]
+     *  :arg point: ``Array`` point to project
+     *  :arg map: ``Object`` the map object
+     *
+     *  It projects the point using the projection list 
+     *  and finds the first one for which it falls
+     *  inside of the restricted Extent.
+     */
+    tryProjection: function(point, map) {
```
Comment (sbrunner) [inline openlayers.addins/AutoProjection/lib/OpenLayers/Util/AutoProjection.js]: Can't we specify the map in the constructor ?

File: core/src/script/CGXP/api/Map.js (hunk 15 above)
Comment (gilliM) [inline core/src/script/CGXP/api/Map.js]: I've observed that when the center components are null , it takes the center of the maxExtent, which is not bad. What do you think about it ?

File: openlayers.addins/AutoProjection/lib/OpenLayers/Util/AutoProjection.js [hunk 22]
```diff
@@ -0,0 +1,86 @@
...
+            code = String(config.projectionCodes[i]).toUpperCase();
+            if (code.substr(0, 5) != "EPSG:") {
+                code = "EPSG:" + code;
+            }
+            this.projections[i] = new OpenLayers.Projection(code);
+        }
+        return;
+    },
+
+    /** private: method[tryProjection]
```
Comment (sbrunner) [inline openlayers.addins/AutoProjection/lib/OpenLayers/Util/AutoProjection.js]: private => public

File: core/src/script/CGXP/api/Map.js (hunk 15 above)
Comment (sbrunner) [inline core/src/script/CGXP/api/Map.js]: Looks good, than:
```
var centerLonLat = this.autoProjection.tryProjection(center, this.map);
if (centerLonLat === undefined) {
  centerLonLat = new OpenLayers.LonLat(null, null);
}
this.map.setCenter(centerLonLat, zoom);
}
```

**Pull request merged**